concurrency/JobStack.java

```diff
@@ -0,0 +1,157 @@
+        }
+    }
+
+    //private LinkedBlockingDeque<JobContext<T>> mTaskQueue = new LinkedBlockingDeque<JobContext<T>>();
+    private IContextQueue<T> mTaskQueue = new LBDContextQueue<T>();
+    private HashMap<String, JobContext<T>> mTaskMap = new HashMap<String, JobContext<T>>();
+
+    /**
+     * This method is used to put job requests into stack.
+     *
+     * @param metadata the metadata for the job
+     * @param listener the listener to notify when job state changes
+     * @return the context of the job
+     */
+    public JobContext<T> put(T metadata, JobListener<T> listener) {
+        JobContext<T> imageLoadContext = null;
+        String key = metadata.getKey();
+        if (key != null) {
+            synchronized (mTaskMap) {
+                if (mTaskMap.containsKey(key)) {
+                    imageLoadContext = mTaskMap.get(key);
+                    imageLoadContext.addListener(listener);
+                    if (!metadata.isPrecacheRequest()) {
+                        imageLoadContext.getMetadata().setPrecacheRequest(false);
+                    }
+                    if (!metadata.isLowPriority()) {
+                        imageLoadContext.getMetadata().setLowPriority(false);
+                        if (mTaskQueue.remove(imageLoadContext)) {
+                            mTaskQueue.addFirst(imageLoadContext);
+                        }
+                    }
+                } else if (!metadata.isUpdateOnly()) {
+                    imageLoadContext = new JobContext<T>(metadata);
+                    imageLoadContext.addListener(listener);
+                    mTaskMap.put(key, imageLoadContext);
+                    if (metadata.isLowPriority()) {
+                        mTaskQueue.addLast(imageLoadContext);
+                    } else {
+                        mTaskQueue.addFirst(imageLoadContext);
+                    }
+                }
+            }
+        }
+        return imageLoadContext;
+    }
+
+    /**
+     * This method will provide the next job to be executed. The
+     * call to this method will block until a job is available.
+     *
+     * @return the next job to be executed
+     */
+    public JobContext<T> take() throws InterruptedException {
+        return mTaskQueue.take();
+    }
+
+    /**
+     * Marks the job represented by the give job context as complete.
+     *
+     * @param jobContext the job context
+     */
+    public void markComplete(JobContext<T> jobContext) {
+        synchronized (mTaskMap) {
+            mTaskMap.remove(jobContext.getMetadata().getKey());
+        }
+    }
+
+    /**
+     * Checks if the stack is empty or not.
+     *
+     * @return true if the stack is empty.
+     */
+    public boolean isEmpty() {
+        synchronized (mTaskMap) {
+            return mTaskQueue.isEmpty();
+        }
+    }
+
+    /**
+     * Empty out the job stack
+     */
+    public void clearStack() {
+        mTaskQueue.clear();
+        mTaskMap.clear();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **144**. At this line the following field is not protected by synchronization: `mTaskMap`. Your code typically protects this field with synchronization on object: `mTaskMap`.

**Fix**  
 Consider protecting the above field with synchronization on object: `mTaskMap`. For example, like at lines: **124**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **80**, **81**, **95**, and **124**.